ENH: subclass-friendly refactor RegularGridInterpolator

kmuehlbauer opened the pull request:
<!-- 
Thanks for contributing a pull request! Please ensure that
your PR satisfies the checklist before submitting:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#checklist-before-submitting-a-pr

Also, please name and describe your PR as you would write a
commit message:
http://scipy.github.io/devdocs/dev/contributor/development_workflow.html#writing-the-commit-message

Note that we are a team of volunteers; we appreciate your
patience during the review process.

Again, thanks for contributing!
-->

#### Reference issue
<!--Example: Closes gh-WXYZ.-->
Supersedes gh-10860
#### What does this implement/fix?
<!--Please explain your changes.-->

Code is refactored to be easier to subclass (eg. overriding methods)   

#### Additional information
<!--Any additional information you think is important.-->

Added one instance variable `_descending_dimensions` which holds a `tuple` of `int` which hold the dimensions with descending order. `points` and `values` are flipped according to that tuple. Removed the current (very slow) code to sort `points`. 

A subclass where `values` instead of `xi` are evaluated on `__call__` would look like this:

```python
class RegularGridInterpolatorValues(RegularGridInterpolator):
    def __init__(self, points, xi, **kwargs):
        # create fake values for initialization
        values = np.zeros(tuple([len(pt) for pt in points]))
        super().__init__(points, values, **kwargs)
        self._is_initialized = False
        # precompute values
        self.xi, self.xi_shape, self.ndim, self.nans, self.out_of_bounds = self._prepare_xi(xi)
        self.indices, self.norm_distances = self._find_indices(xi.T)
        self._is_initialized = True

    def _prepare_xi(self, xi):
        if not self._is_initialized:
            return super()._prepare_xi(xi)
        else:
            # just give back precomputed values
            return self.xi, self.xi_shape, self.ndim, self.nans, self.out_of_bounds

    def _find_indices(self, xi):
        if not self._is_initialized:
            return super()._find_indices(xi)
        else:
            # just give back pre-computed values
            return self.indices, self.norm_distances

    def __call__(self, values, method=None):
        valiues = self._check_values(values)
        # check fillvalue
        self._check_fill_value(values, self.fill_value)
        # check dimensionality
        self._check_dimensionality(self.grid, values)
        # flip, if needed
        self.values = np.flip(values, axis=self.flip)
        return super().__call__(self.xi, method=method)
```

kmuehlbauer commented:
@ev-br Here is the WIP PullRequest. Please be patient with me, I've had to refactor a lot of the code to better be able to subclass it. I'm sure you have some valuable advise how to do better. 

Trying to fix linting now.

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,20 +217,41 @@ def __init__(self, points, values, method="linear", bounds_error=True,
             self._validate_grid_dimensions(points, method)
         self.method = method
         self.bounds_error = bounds_error
+        # todo: sort checks in order of complexity and runtime, raising early
```
Minor:This is something you are planning to do in this PR, or is it done and the comment should be removed already?

kmuehlbauer replied:
Done

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,20 +217,41 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+        self._check_fill_value(values, fill_value)
+        self.fill_value = fill_value
+        flip = self._check_points(points)
+        self.flip = flip
```
`self.flip` should probably be private, i.e. should have a leading underscore. We are trying to minimize the API surface. Also this variable, or the function which computes it, needs a comment for future readers.

kmuehlbauer replied:
Renamed to a more descriptive name and made private. Would that be enough? If not where should I put that comment? Above the call to _check_points or inside that function?

ev-br reviewed (commented):
Looks reasonable on a cursory look, will review in more details a bit later. Meanwhile, can you show the profile and timings compared to main? Eg, the same synthetic test case you had in the original PR?

ev-br commented:
I'd ask @j-bowhay and @AtsushiSakai to have a look, as you two worked with this code recently, would you?

kmuehlbauer commented:
@ev-br I've added changes with regard to your first review.

Here is a gist with the profiling on current main:

https://gist.github.com/kmuehlbauer/bf3d2d16c7a3fc592b4c8fd8d33d7416

And here is one gist with the profiling using that PR branch:

https://gist.github.com/kmuehlbauer/81b99492c964a67912cd7910cde8f816

ev-br commented:
Two comments. First, I have to admit I'm confused by your statements about timing. Your original profile was hinting on some 2-3x improvement, but you say it's only 20%. Then there is a statement about something being several orders of magnitude faster/slower, and I'm failing to see it. Most likely I'm just being dense.

It would be useful to add an asv benchmark (https://github.com/scipy/scipy/tree/main/benchmarks) using a maybe better test case, like you mentioned. Then it would help to add as a comment the result of a `--bench-compare` of this PR vs main. For both RGI and your subclass. Maybe best benchmark both construction and evaluation, both are useful here and will be useful for gh-16483.

Second, ascending/descending/unsorted points. Strictly speaking, the argsorting logic was released in scipy 1.9.0, so it is technically a backcompat break to change that. OTOH, we did not advertise a full `assume_sorted=False` type functionality, so maybe, just maybe, we can change that for 1.10. If there is a compelling reason for it. To decide on this, I'm going to defer to @AtsushiSakai .

Also to the flipping logic, I'm not sure how this handles trailing dimensions of `values` etc. I'm not actually sure we have good enough test coverage for trailing dimensions *and* descending/unsorted input. Which is one reason why I pinged Jake Bowhay above.

And, finally, feel free to ignore the CirrusCI failures, they seem unrelated.

kmuehlbauer commented:
> First, I have to admit I'm confused by your statements about timing. Your original profile was hinting on some 2-3x improvement, but you say it's only 20%. Then there is a statement about something being several orders of magnitude faster/slower, and I'm failing to see it. Most likely I'm just being dense.

@ev-br For the timing, 20% means the new code takes only 20% of the time the other code took. So from say 100s to 20s. Does that make sense?

The orders of magnitude refer to the example if descending coordinates are present. In my example just one run already took 1.7 seconds instead of 0.005 seconds if no descending coordinates are present. That's quite something (340x). 

> Second, ascending/descending/unsorted points. Strictly speaking, the argsorting logic was released in scipy 1.9.0, so it is technically a backcompat break to change that. OTOH, we did not advertise a full assume_sorted=False type functionality, so maybe, just maybe, we can change that for 1.10. If there is a compelling reason for it. To decide on this, I'm going to defer to @AtsushiSakai .

From my understanding, the current code and my implementation are equivalent with regard to functionality. Why I'm thinking that? Because `_make_points_and_values_ascending` is only called when a strictly descending dimension is detected. So this is just a simple flip that code has to do:

https://github.com/scipy/scipy/blob/659975ffef30af0ef9e0902d2db4bf4f149185f9/scipy/interpolate/_rgi.py#L242-L252

That means currently only either strictly ascending or strictly descending dimensions are allowed, any other setup will raise an error.

> Also to the flipping logic, I'm not sure how this handles trailing dimensions of values etc. I'm not actually sure we have good enough test coverage for trailing dimensions and descending/unsorted input. Which is one reason why I pinged Jake Bowhay above.

I'm sure I'm missing something here. In the proposed implementation the descending dimensions are detected and the according `points` are flipped. Additionally `np.flip` is used and those dimensions are flipped for `values` accordingly. Can you explain what trailing dimensions are in this context? 


> It would be useful to add an asv benchmark (https://github.com/scipy/scipy/tree/main/benchmarks) using a maybe better test case, like you mentioned. Then it would help to add as a comment here the result of a --bench-compare of this PR vs main. For both RGI and your subclass. Maybe best benchmark both construction and evaluation, both are useful here and will be useful for https://github.com/scipy/scipy/pull/16483.


The benchmark add-on will take some time. I'm not versed with that stuff and will have to dive into first.

ev-br commented:
> For the timing, 20% means the new code takes only 20% of the time the other code took. So from say 100s to 20s. Does that make sense?

Ah, so a 5x improvement. Nice!
Ditto for the sorting improvement.

> That means currently only either strictly ascending or strictly descending dimensions are allowed, any other setup will raise an error.

OK, great. No backcompat concerns here then.

> I'm sure I'm missing something here. In the proposed implementation the descending dimensions are detected and the according points are flipped. Additionally np.flip is used and those dimensions are flipped for values accordingly. Can you explain what trailing dimensions are in this context?

It's basically this: https://github.com/scipy/scipy/blob/main/scipy/interpolate/tests/test_rgi.py#L511
Note that for an N-D problem `values` are allowed to have more than N dimensions, and evaluations are vectorized across them.
I'm not saying that your refactor is incorrect! Most likely it is. All I'm saying flipping is not tested for values with trailing axes, and it should be. Quite a bit of work went into RGI code recently just because things were undertested and thus subtly wrong in some weird ways (your earlier comment that internals have changed, that's it).
So it might be as simple as adding a test where some axes need to be flipped and some not, and checking that the result is identical to when the points are pre-sorted.

> The benchmark add-on will take some time. I'm not versed with that stuff and will have to dive into first.

Yes. Luckily, we're not on a deadline.
However fiddly, asv benchmarks do add value. Both during the PR time (via a standartized way of comparing a feature branch to main), and long-term, as a standardized way of checking for perf regressions or improvement.

kmuehlbauer commented:
@ev-br Thanks for the details. I'll see that I get into benchmarking as long as memory is fresh. I'll also add a test for the flipping stuff. Should it go where you linked?

ev-br commented:
> Should it go where you linked?

Why not. A separate test just next to, or something such.

kmuehlbauer commented:
@ev-br I've parameterized that test you linked to also test for all strictly decreasing points dimensions. Works locally, so suspect to get green tests here too. Now looking into the benchmarking.

kmuehlbauer commented:
OK, here is some benchmarking. I've added the same parameterizing as in the test.

I've added a benchmark just for the interpolator creation. So, I think this is shows a good performance, especially for the descending dimensions.

<pre>[100.00%] ···· <font color="#CC0000">For parameters: 4, 200, 10, -1</font>
               
               
               asv: benchmark timed out (timeout 60.0s)
               
               For parameters: 4, 200, 100, -1
               
               
               asv: benchmark timed out (timeout 60.0s)
               
               For parameters: 4, 200, 1000, -1
               
               
               asv: benchmark timed out (timeout 60.0s)
               
               For parameters: 4, 200, 10000, -1
               
               
               asv: benchmark timed out (timeout 60.0s)

       before           after         ratio
     [5d7b60a5]       [d113c843]
     &lt;main&gt;           &lt;rgi&gt;     
<font color="#CC0000">+      29.8±0.9μs         74.2±4μs     2.49  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10000, 1)
<font color="#CC0000">+      29.8±0.8μs         74.2±4μs     2.49  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 1000, 1)
<font color="#CC0000">+      23.1±0.2μs         28.8±2μs     1.25  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 1000, 1)
<font color="#CC0000">+      21.9±0.7μs         26.6±1μs     1.21  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10, 1)
<font color="#CC0000">+      22.3±0.8μs         26.9±1μs     1.21  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10, 1)
<font color="#CC0000">+      22.9±0.6μs         27.4±2μs     1.20  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10, 1)
<font color="#CC0000">+      22.2±0.6μs         26.7±1μs     1.20  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10000, 1)
<font color="#CC0000">+      22.7±0.1μs         27.1±1μs     1.19  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 100, 1)
<font color="#CC0000">+      22.6±0.8μs         26.7±1μs     1.19  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 1000, 1)
<font color="#CC0000">+        30.0±1μs         35.4±2μs     1.18  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 100, 1)
<font color="#CC0000">+      22.5±0.9μs         26.5±1μs     1.18  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 100, 1)
<font color="#CC0000">+        29.7±1μs         34.6±2μs     1.16  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10, 1)
<font color="#CC0000">+      29.3±0.6μs         34.1±2μs     1.16  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 100, 1)
<font color="#CC0000">+        30.9±1μs         35.9±2μs     1.16  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 10, 1)
<font color="#CC0000">+        31.3±1μs         36.3±2μs     1.16  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 100, 1)
<font color="#CC0000">+        31.2±1μs         35.9±2μs     1.15  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 1000, 1)
<font color="#CC0000">+        38.9±1μs         44.5±2μs     1.14  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 200, 1000, 1)
<font color="#CC0000">+      30.1±0.9μs         34.3±2μs     1.14  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 1000, 1)
<font color="#CC0000">+        30.4±1μs         34.6±2μs     1.14  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10000, 1)
<font color="#CC0000">+      38.4±0.7μs         42.9±2μs     1.12  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 10000, 1)
<font color="#4E9A06">-         619±9μs          582±6μs     0.94  </font>interpolate.RegularGridInterpolator.time_rgi(4, 40, 1000, -1)
<font color="#4E9A06">-         135±3μs          126±4μs     0.93  </font>interpolate.RegularGridInterpolator.time_rgi(2, 40, 100, 1)
<font color="#4E9A06">-     1.16±0.02ms         1.08±0ms     0.93  </font>interpolate.RegularGridInterpolator.time_rgi(2, 40, 10000, 1)
<font color="#4E9A06">-     1.95±0.06ms      1.80±0.01ms     0.92  </font>interpolate.RegularGridInterpolator.time_rgi(3, 40, 10000, -1)
<font color="#4E9A06">-         251±8μs          229±3μs     0.92  </font>interpolate.RegularGridInterpolator.time_rgi(2, 40, 1000, -1)
<font color="#4E9A06">-     1.53±0.07ms         1.39±0ms     0.91  </font>interpolate.RegularGridInterpolator.time_rgi(2, 200, 10000, -1)
<font color="#4E9A06">-        401±10μs          363±4μs     0.90  </font>interpolate.RegularGridInterpolator.time_rgi(3, 40, 1000, -1)
<font color="#4E9A06">-         277±7μs          250±4μs     0.90  </font>interpolate.RegularGridInterpolator.time_rgi(4, 40, 10, -1)
<font color="#4E9A06">-         144±6μs          130±4μs     0.90  </font>interpolate.RegularGridInterpolator.time_rgi(2, 200, 100, -1)
<font color="#4E9A06">-         128±5μs          114±2μs     0.89  </font>interpolate.RegularGridInterpolator.time_rgi(2, 200, 10, 1)
<font color="#4E9A06">-        299±20μs          266±2μs     0.89  </font>interpolate.RegularGridInterpolator.time_rgi(2, 200, 1000, -1)
<font color="#4E9A06">-         131±5μs          116±3μs     0.89  </font>interpolate.RegularGridInterpolator.time_rgi(2, 200, 10, -1)
<font color="#4E9A06">-         140±5μs          124±2μs     0.88  </font>interpolate.RegularGridInterpolator.time_rgi(2, 40, 100, -1)
<font color="#4E9A06">-       536±200μs          421±7μs     0.79  </font>interpolate.RegularGridInterpolator.time_rgi(3, 200, 1000, 1)
<font color="#4E9A06">-      1.32±0.4ms          821±4μs     0.62  </font>interpolate.RegularGridInterpolator.time_rgi(2, 10, 10000, -1)
<font color="#4E9A06">-      1.32±0.4ms          818±5μs     0.62  </font>interpolate.RegularGridInterpolator.time_rgi(2, 10, 10000, 1)
<font color="#4E9A06">-        199±70μs          117±3μs     0.59  </font>interpolate.RegularGridInterpolator.time_rgi(2, 40, 10, -1)
<font color="#4E9A06">-        927±20μs          111±5μs     0.12  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 1000, -1)
<font color="#4E9A06">-        941±20μs          112±4μs     0.12  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10000, -1)
<font color="#4E9A06">-         931±5μs          109±6μs     0.12  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 100, -1)
<font color="#4E9A06">-         374±8μs         38.6±2μs     0.10  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10000, -1)
<font color="#4E9A06">-         372±8μs         38.3±2μs     0.10  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 1000, -1)
<font color="#4E9A06">-         371±3μs         38.1±2μs     0.10  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 100, -1)
<font color="#4E9A06">-         375±6μs         38.2±1μs     0.10  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10, -1)
<font color="#4E9A06">-     1.04±0.02ms         59.1±3μs     0.06  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 1000, -1)
<font color="#4E9A06">-     1.04±0.03ms         58.9±3μs     0.06  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 100, -1)
<font color="#4E9A06">-     1.05±0.02ms         59.0±3μs     0.06  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 10000, -1)
<font color="#4E9A06">-        928±30μs         51.6±2μs     0.06  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10, -1)
<font color="#4E9A06">-     1.06±0.02ms         58.7±4μs     0.06  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 10, -1)
<font color="#4E9A06">-     3.03±0.04ms         38.2±2μs     0.01  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10000, -1)
<font color="#4E9A06">-     3.09±0.06ms         38.3±2μs     0.01  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10, -1)
<font color="#4E9A06">-        4.61±2ms         37.9±2μs     0.01  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 1000, -1)
<font color="#4E9A06">-        4.71±2ms         38.3±2μs     0.01  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 100, -1)
<font color="#4E9A06">-      47.5±0.8ms         51.5±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10000, -1)
<font color="#4E9A06">-      47.6±0.4ms         51.2±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 1000, -1)
<font color="#4E9A06">-      47.5±0.6ms         50.7±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 100, -1)
<font color="#4E9A06">-       58.1±20ms         50.4±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10, -1)
<font color="#4E9A06">-      72.7±0.2ms         39.7±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 100, -1)
<font color="#4E9A06">-      72.4±0.8ms         39.4±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10, -1)
<font color="#4E9A06">-        73.2±2ms         39.7±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10000, -1)
<font color="#4E9A06">-      73.1±0.3ms         39.2±2μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 1000, -1)
<font color="#4E9A06">-         340±8ms         67.4±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 1000, -1)
<font color="#4E9A06">-         342±3ms         66.1±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 100, -1)
<font color="#4E9A06">-         339±4ms         65.5±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 10000, -1)
<font color="#4E9A06">-         339±7ms         64.8±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 10, -1)
<font color="#4E9A06">-      6.27±0.03s         53.0±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 10, -1)
<font color="#4E9A06">-      6.47±0.06s         52.9±3μs     0.00  </font>interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 10000, -1)
</pre>

ev-br commented:
1. Would be good to have the benchmark committed to the scipy repository.
2. An evaluation benchmark would be nice, too. 
3. The test you added needs some work, I'm afraid. I think a new test should compare two interpolations, one with all in order and the other with all or some dimensions flipped. And the test checks that the results match.

kmuehlbauer commented:
> 3\. The test you added needs some work, I'm afraid. I think a new test should compare two interpolations, one with all in order and the other with all or some dimensions flipped. And the test checks that the results match.

I think this is checked already here: https://github.com/scipy/scipy/blob/f5508d8f10f7e6e649059f91bb0de94ae682cdea/scipy/interpolate/tests/test_rgi.py#L437

I'll enhance that test to cover all methods, to be on the safe side.

kmuehlbauer commented:
@ev-br I've added a benchmark test to the existing class and duplicated that to work on our  subclass.

So here is the benchmark just for the current branch:

```
### -> this times only the evaluation (__call__)
[ 88.24%] ··· interpolate.RegularGridInterpolator.time_rgi                                                                                                                                                       ok
[ 88.24%] ··· ====== ================ ========= =========== ========== ========== ========== =========== ============= =============
              --                                                           n_samples / flipped                                      
              ----------------------- ----------------------------------------------------------------------------------------------
               ndim   max_coord_size    10 / 1    10 / -1    100 / 1    100 / -1   1000 / 1   1000 / -1    10000 / 1     10000 / -1 
              ====== ================ ========= =========== ========== ========== ========== =========== ============= =============
                2           10         160±2μs    154±6μs    165±5μs    188±20μs   230±3μs     234±10μs     804±5μs       840±30μs  
                2           40         157±6μs   162±0.6μs   165±5μs    165±4μs    261±9μs     262±7μs    1.11±0.05ms   1.09±0.02ms 
                2          200         159±6μs    158±4μs    169±5μs    172±7μs    295±8μs     298±10μs   1.41±0.04ms   1.45±0.04ms 
                3           10         222±4μs    226±2μs    243±10μs   244±7μs    357±10μs    353±10μs   1.33±0.04ms   1.36±0.04ms 
                3           40         227±2μs    232±6μs    245±10μs   236±5μs    395±20μs    401±10μs   1.78±0.09ms   1.77±0.07ms 
                3          200         238±5μs    229±2μs    253±7μs    254±2μs    457±8μs     455±5μs     2.60±0.1ms   2.59±0.05ms 
                4           10         332±5μs    334±6μs    353±7μs    372±2μs    571±10μs    548±7μs     2.39±0.1ms   2.41±0.06ms 
                4           40         335±5μs    341±8μs    358±5μs    358±10μs   630±7μs     626±8μs    2.84±0.03ms   2.79±0.07ms 
                4          200         339±6μs    340±2μs    379±10μs   372±7μs    712±8μs     707±10μs    6.31±0.2ms    6.58±0.2ms 
              ====== ================ ========= =========== ========== ========== ========== =========== ============= =============

### -> this times only the instatiation
[ 91.18%] ··· interpolate.RegularGridInterpolator.time_rgi_setup_interpolator                                                                                                                                    ok
[ 91.18%] ··· ====== ================ ============ ============ ============ ============ ============ ============ ============ ============
              --                                                                n_samples / flipped                                          
              ----------------------- -------------------------------------------------------------------------------------------------------
               ndim   max_coord_size     10 / 1      10 / -1      100 / 1      100 / -1     1000 / 1    1000 / -1    10000 / 1    10000 / -1 
              ====== ================ ============ ============ ============ ============ ============ ============ ============ ============
                2           10         38.0±0.7μs    54.9±1μs    36.5±0.4μs   51.0±0.3μs    37.0±1μs    53.7±0.9μs   38.4±0.7μs    52.9±1μs  
                2           40         37.4±0.5μs   51.6±0.9μs   36.8±0.3μs    52.2±1μs    37.6±0.7μs   52.1±0.7μs    36.8±1μs     51.6±3μs  
                2          200          37.4±2μs     53.9±2μs    45.7±10μs     53.7±3μs     38.6±1μs    53.9±0.8μs   38.9±0.6μs    52.9±1μs  
                3           10          48.5±1μs     70.0±1μs    48.5±0.7μs    71.0±1μs     48.1±1μs     70.8±2μs     48.4±1μs     69.6±2μs  
                3           40         56.6±10μs    86.6±20μs    47.8±10μs     69.3±2μs    47.2±0.4μs   70.6±0.8μs   47.3±0.7μs   67.7±0.6μs 
                3          200         47.8±0.3μs   69.7±0.8μs   48.0±0.7μs   68.4±0.8μs   47.2±0.3μs   69.1±0.4μs   48.1±0.8μs    69.3±1μs  
                4           10          56.6±1μs     78.7±1μs     57.7±1μs    79.1±0.6μs   56.6±0.5μs    79.9±1μs    56.6±0.5μs   79.1±0.7μs 
                4           40         58.9±0.6μs    86.7±3μs     58.7±2μs     87.3±2μs    59.6±0.9μs    87.6±2μs     58.5±2μs     87.1±2μs  
                4          200          58.5±1μs     86.0±1μs     59.0±2μs     88.6±2μs    57.8±0.5μs    86.1±2μs    55.8±0.6μs   84.5±0.8μs 
              ====== ================ ============ ============ ============ ============ ============ ============ ============ ============

### -> this times only the evaluation (__call__) (SubClass)
[ 94.12%] ··· interpolate.RegularGridInterpolatorSubclass.time_rgi                                                                                                                                               ok
[ 94.12%] ··· ====== ================ ============ ========== ============ ============ ========== ============ ============= =============
              --                                                               n_samples / flipped                                         
              ----------------------- -----------------------------------------------------------------------------------------------------
               ndim   max_coord_size     10 / 1     10 / -1     100 / 1      100 / -1    1000 / 1   1000 / -1     10000 / 1     10000 / -1 
              ====== ================ ============ ========== ============ ============ ========== ============ ============= =============
                2           10         64.6±10μs    67.6±2μs    66.9±2μs     68.5±1μs    85.5±1μs   87.8±0.4μs     267±3μs       260±4μs   
                2           40         61.6±0.6μs   64.7±2μs   62.8±0.6μs   68.5±0.7μs   88.8±2μs    86.5±2μs      271±9μs       270±4μs   
                2          200         63.6±0.8μs   65.0±2μs    64.5±1μs     67.5±1μs    88.2±1μs    93.7±2μs      284±9μs       289±8μs   
                3           10          98.1±1μs    97.0±2μs   104±0.7μs     105±3μs     154±4μs     157±4μs       583±9μs       608±20μs  
                3           40         96.5±0.7μs   99.8±2μs    103±2μs     102±0.9μs    159±4μs     154±2μs       603±7μs       600±10μs  
                3          200          98.0±1μs    97.5±1μs    104±2μs      104±1μs     166±4μs     168±3μs      866±200μs      765±10μs  
                4           10          160±2μs     163±2μs     174±3μs      176±2μs     291±2μs     295±5μs     1.32±0.06ms   1.30±0.04ms 
                4           40          162±6μs     170±2μs     179±1μs      178±4μs     296±2μs     301±3μs     1.42±0.02ms   1.45±0.02ms 
                4          200          162±2μs     169±3μs     181±7μs      182±4μs     329±8μs     336±10μs     3.80±0.3ms    4.15±0.2ms 
              ====== ================ ============ ========== ============ ============ ========== ============ ============= =============

### -> this times only the instatiation (Subclass)
[ 97.06%] ··· interpolate.RegularGridInterpolatorSubclass.time_rgi_setup_interpolator                                                                                                                                                                                       ok
[ 97.06%] ··· ====== ================ ============= ============= ============= ============= ============= ============= ============= =============
              --                                                                    n_samples / flipped                                              
              ----------------------- ---------------------------------------------------------------------------------------------------------------
               ndim   max_coord_size      10 / 1       10 / -1       100 / 1       100 / -1      1000 / 1     1000 / -1     10000 / 1     10000 / -1 
              ====== ================ ============= ============= ============= ============= ============= ============= ============= =============
                2           10           156±4μs       208±50μs      218±60μs      199±30μs      252±60μs      229±4μs       655±70μs      646±20μs  
                2           40           159±3μs       174±3μs       162±5μs       178±3μs       231±3μs       256±3μs       912±9μs       918±10μs  
                2          200           162±3μs      175±0.3μs      170±3μs       182±2μs       270±4μs       285±3μs     1.24±0.04ms   1.27±0.03ms 
                3           10           210±6μs       233±4μs       214±2μs       239±5μs       278±4μs       295±3μs       838±10μs      864±8μs   
                3           40           231±30μs      250±60μs      224±6μs       249±2μs       332±3μs       356±4μs      1.31±0.2ms   1.32±0.02ms 
                3          200         1.11±0.03ms   1.09±0.02ms   1.06±0.03ms   1.15±0.04ms   1.26±0.08ms   1.27±0.05ms   2.70±0.07ms    2.74±0.2ms 
                4           10           266±5μs       283±8μs       270±7μs       297±5μs       345±8μs       379±5μs     1.01±0.01ms   1.04±0.05ms 
                4           40          335±100μs     369±100μs      284±70μs      312±10μs      413±5μs       442±5μs     1.48±0.03ms   1.49±0.03ms 
                4          200           272±2μs       304±6μs       287±7μs       322±6μs       473±10μs      502±5μs     2.11±0.03ms   2.19±0.05ms 
              ====== ================ ============= ============= ============= ============= ============= ============= ============= =============

```

So the evaluation times are a factor of 4 to 5 better using the prepared sublass.

For one cycle (instantiation + evaluation) both methods are comparable. But for the subclass use case it will play out the better evaluation times when called multiple times in a row.

Here is the benchmark comparison to main branch:

```
+      29.6±0.4μs       35.9±0.9μs     1.21  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 100, 1)
+      30.8±0.2μs         37.4±1μs     1.21  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 1000, 1)
+      29.3±0.3μs       35.5±0.6μs     1.21  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 1000, 1)
+      30.8±0.6μs       37.2±0.2μs     1.21  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 100, 1)
+        39.6±1μs       47.4±0.5μs     1.20  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 1000, 1)
+      30.1±0.5μs       35.8±0.5μs     1.19  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10, 1)
+      31.0±0.3μs         36.4±1μs     1.18  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10, 1)
+      29.8±0.8μs       35.0±0.3μs     1.18  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10000, 1)
+      30.6±0.3μs       35.9±0.4μs     1.17  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10, 1)
+      31.1±0.5μs       36.0±0.6μs     1.16  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10000, 1)
+      31.5±0.3μs         36.4±1μs     1.15  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10000, 1)
+      30.6±0.2μs       35.2±0.4μs     1.15  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 100, 1)
+      40.3±0.7μs         46.4±2μs     1.15  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 100, 1)
+        40.3±1μs       46.1±0.7μs     1.14  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 100, 1)
+      40.0±0.5μs         45.7±1μs     1.14  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10000, 1)
+      40.3±0.7μs         46.0±1μs     1.14  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 100, 1)
+      39.9±0.5μs         45.2±2μs     1.13  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10, 1)
+      41.6±0.4μs         46.8±1μs     1.13  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10, 1)
+        40.2±1μs       45.2±0.7μs     1.12  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 10000, 1)
+        41.3±1μs         46.3±1μs     1.12  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 10, 1)
+        40.6±1μs       45.3±0.4μs     1.12  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 1000, 1)
+        49.9±1μs       55.6±0.8μs     1.11  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 100, 1)
+      49.5±0.6μs       54.7±0.7μs     1.11  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 1000, 1)
+        49.9±1μs       54.7±0.8μs     1.10  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 10, 1)
+         246±3μs          270±2μs     1.09  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 100, 1)
+      41.5±0.5μs       45.0±0.3μs     1.08  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10000, 1)
+      50.5±0.9μs       54.6±0.4μs     1.08  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 1000, 1)
+      42.5±0.5μs       45.6±0.6μs     1.07  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 1000, 1)
+     6.40±0.05ms       6.86±0.1ms     1.07  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 200, 100, 1)
-        461±20μs       51.7±0.8μs     0.11  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 1000, -1)
-        461±20μs       51.5±0.7μs     0.11  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10, -1)
-         459±9μs       50.0±0.7μs     0.11  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 10000, -1)
-         459±8μs       49.6±0.9μs     0.11  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 10, 100, -1)
-     1.24±0.01ms       76.5±0.8μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 100, -1)
-     1.24±0.01ms       76.8±0.5μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 10, -1)
-     1.23±0.01ms         75.9±2μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 10000, -1)
-     1.14±0.02ms         69.3±1μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 100, -1)
-     1.16±0.02ms         69.9±1μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10, -1)
-     1.16±0.02ms         69.1±2μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 1000, -1)
-     1.15±0.03ms         68.6±1μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 10, 10000, -1)
-     1.29±0.04ms         75.8±1μs     0.06  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 10, 1000, -1)
-     4.09±0.03ms         52.3±1μs     0.01  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10000, -1)
-     4.19±0.06ms         52.5±1μs     0.01  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 1000, -1)
-     4.12±0.05ms         50.9±1μs     0.01  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 100, -1)
-     4.18±0.02ms       51.0±0.9μs     0.01  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 40, 10, -1)
-        61.0±1ms         70.7±5μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 100, -1)
-      61.2±0.7ms         69.9±1μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 1000, -1)
-      61.2±0.8ms         69.6±2μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10, -1)
-        61.6±1ms         65.4±2μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 40, 10000, -1)
-      95.0±0.3ms         54.0±4μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10000, -1)
-        96.3±1ms       52.9±0.7μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 10, -1)
-      96.2±0.6ms         52.2±1μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 1000, -1)
-      96.1±0.8ms       51.9±0.9μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(2, 200, 100, -1)
-         424±6ms         85.7±2μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 100, -1)
-         426±6ms         82.9±1μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 10000, -1)
-         427±8ms         82.5±1μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 10, -1)
-         426±8ms       81.3±0.5μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(4, 40, 1000, -1)
-         7.24±0s         67.1±1μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 10000, -1)
-      7.35±0.02s       67.2±0.8μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 1000, -1)
-      7.31±0.05s         66.6±1μs     0.00  interpolate.RegularGridInterpolator.time_rgi_setup_interpolator(3, 200, 100, -1)

```

This can't compare the Subclass, so only normal `RegularGridInterpolator` is shown. There slightly increasing values for the normal cases, most likely due to the function overheads (?). Maybe we can do a bit better here. But for the decreasing dimensions the new implementation is way better.

ev-br commented:
It would be nice to check the profile for the small size problems where benchmarks report non-trivial slowdowns. If it's just a function call overheads, there is not much to be done indeed. But maybe it can be improved.

Overall this PR looks good, am marking it for the 1.10 release so that it does not fall through the cracks.
I'll keep this open for now so that others have time to have a look.
(Atsushi Sakai mentioned elsewhere that he's busy these days at least).

An obligatory note about subclassing. Am adding to put it on record. The intended subclass reaches into the private internals of the RGI class (those with a leading underscore). Strictly speaking, they are not covered by the backwards compatibility guarantees, and may change in future scipy versions. While the chances of significant changes are rather low, some downstream maintenance of the subclass usage may be required going forward.

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,20 +217,39 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+        values = self._check_values(values)
+        self._check_dimensionality(points, values)
+        self._check_fill_value(values, fill_value)
+        self.fill_value = fill_value
```
Minor: any reason to not have `self.fill_value = self._check_fill_value(fill_value)`

kmuehlbauer replied:
Applied.

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,20 +217,39 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+        self._descending_dimensions = self._check_points(points)
+        self.grid = tuple([np.flip(p) if i in self._descending_dimensions
+                           else np.asarray(p)
+                           for i, p in enumerate(points)])
```
gh-16483 needs to make sure that the grid points are floats: `self.grid = tuple([np.asarray(p, dtype=float) for p in points])`, could make it like this here already.

kmuehlbauer replied:
I've took the chance to restructure and possibly make this a bite more readable.

ev-br reviewed (commented):
Two minor suggestions, both optional. Are both only worth doing if there is something to lump them to. 

A bigger question is whether this is anything to fix in the initialization where benchmarks show slow-downs. Here a line profile could be helpful.

kmuehlbauer commented:
> It would be nice to check the profile for the small size problems where benchmarks report non-trivial slowdowns. If it's just a function call overheads, there is not much to be done indeed. But maybe it can be improved.

I've tinkered a bit with this and switched the `np.diff`-based check for monotonicity to a simple compare with neighbour. This alone reduces the overall `__init__` runtime to half.

> Overall this PR looks good, am marking it for the 1.10 release so that it does not fall through the cracks. I'll keep this open for now so that others have time to have a look. (Atsushi Sakai mentioned elsewhere that he's busy these days at least).

Great! Yes, I've seen it. Let's just wait a bit then.

> An obligatory note about subclassing. Am adding to put it on record. The intended subclass reaches into the private internals of the RGI class (those with a leading underscore). Strictly speaking, they are not covered by the backwards compatibility guarantees, and may change in future scipy versions. While the chances of significant changes are rather low, some downstream maintenance of the subclass usage may be required going forward.

If you see any move forward to add changes which make subclassing not rely on private internals I'll gladly add this. As I see this, we would need to add the `_check_something` to user API as well as `_prepare_xi` and `_find_indices`. Not sure, this is wanted. Maybe it's best to closely follow scipy development and keep in sync that way.

kmuehlbauer commented:
Line Profiling `__init__` with latest commit in this PR:

```
Timer unit: 1e-06 s

Total time: 0.055326 s
File: /home/kai/miniconda/envs/scipy-dev/lib/python3.10/site-packages/scipy/interpolate/_rgi.py
Function: __init__ at line 212

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   212                                               def __init__(self, points, values, method="linear", bounds_error=True,
   213                                                            fill_value=np.nan):
   214      1000       1294.0      1.3      2.3          if method not in self._ALL_METHODS:
   215                                                       raise ValueError("Method '%s' is not defined" % method)
   216      1000        853.0      0.9      1.5          elif method in self._SPLINE_METHODS:
   217                                                       self._validate_grid_dimensions(points, method)
   218      1000        611.0      0.6      1.1          self.method = method
   219      1000        480.0      0.5      0.9          self.bounds_error = bounds_error
   220      1000      29753.0     29.8     53.8          self.grid, self.descending_dimensions = self._check_points(points)
   221      1000       8257.0      8.3     14.9          self.values = self._check_values(values)
   222      1000       4984.0      5.0      9.0          self._check_dimensionality(self.grid, self.values)
   223      1000       8500.0      8.5     15.4          self.fill_value = self._check_fill_value(self.values, fill_value)
   224      1000        594.0      0.6      1.1          if self.descending_dimensions:
   225                                                       self.values = np.flip(values, axis=self.descending_dimensions)
```

Line Profiling `__init__` with main:

```
Timer unit: 1e-06 s

Total time: 0.100061 s
File: /home/kai/miniconda/envs/scipy-dev/lib/python3.10/site-packages/scipy/interpolate/_rgi.py
Function: __init__ at line 212

Line #      Hits         Time  Per Hit   % Time  Line Contents
==============================================================
   212                                               def __init__(self, points, values, method="linear", bounds_error=True,
   213                                                            fill_value=np.nan):
   214      1000       1675.0      1.7      1.7          if method not in self._ALL_METHODS:
   215                                                       raise ValueError("Method '%s' is not defined" % method)
   216      1000       1122.0      1.1      1.1          elif method in self._SPLINE_METHODS:
   217                                                       self._validate_grid_dimensions(points, method)
   218      1000        937.0      0.9      0.9          self.method = method
   219      1000        723.0      0.7      0.7          self.bounds_error = bounds_error
   220                                           
   221      1000       1038.0      1.0      1.0          if not hasattr(values, 'ndim'):
   222                                                       # allow reasonable duck-typed values
   223                                                       values = np.asarray(values)
   224                                           
   225      1000       1088.0      1.1      1.1          if len(points) > values.ndim:
   226                                                       raise ValueError("There are %d point arrays, but values has %d "
   227                                                                        "dimensions" % (len(points), values.ndim))
   228                                           
   229      1000       1193.0      1.2      1.2          if hasattr(values, 'dtype') and hasattr(values, 'astype'):
   230      1000       6870.0      6.9      6.9              if not np.issubdtype(values.dtype, np.inexact):
   231                                                           values = values.astype(float)
   232                                           
   233      1000        983.0      1.0      1.0          self.fill_value = fill_value
   234      1000        756.0      0.8      0.8          if fill_value is not None:
   235      1000       1969.0      2.0      2.0              fill_value_dtype = np.asarray(fill_value).dtype
   236      2000       1620.0      0.8      1.6              if (hasattr(values, 'dtype') and not
   237      2000       6449.0      3.2      6.4                      np.can_cast(fill_value_dtype, values.dtype,
   238      1000        673.0      0.7      0.7                                  casting='same_kind')):
   239                                                           raise ValueError("fill_value must be either 'None' or "
   240                                                                            "of a type compatible with values")
   241                                           
   242      3000       3176.0      1.1      3.2          for i, p in enumerate(points):
   243      2000      32764.0     16.4     32.7              diff_p = np.diff(p)
   244      2000      28097.0     14.0     28.1              if not np.all(diff_p > 0.):
   245                                                           if np.all(diff_p < 0.):
   246                                                               # input is descending, so make it ascending
   247                                                               points, values = _make_points_and_values_ascending(
   248                                                                   points, values)
   249                                                           else:
   250                                                               raise ValueError(
   251                                                                   "The points in dimension %d must be strictly "
   252                                                                   "ascending or descending" % i)
   253      2000       2436.0      1.2      2.4              if not np.asarray(p).ndim == 1:
   254                                                           raise ValueError("The points in dimension %d must be "
   255                                                                            "1-dimensional" % i)
   256      2000       2988.0      1.5      3.0              if not values.shape[i] == len(p):
   257                                                           raise ValueError("There are %d points and %d values in "
   258                                                                            "dimension %d" % (len(p), values.shape[i], i))
   259      1000       2619.0      2.6      2.6          self.grid = tuple([np.asarray(p) for p in points])
   260      1000        885.0      0.9      0.9          self.values = values
```

j-bowhay commented:
> I'd ask @j-bowhay and @AtsushiSakai to have a look, as you two worked with this code recently, would you?

Hopefully might have time to look at this next week but don't let me hold things up if there is a consensus on this

kmuehlbauer commented:
@ev-br @j-bowhay @AtsushiSakai 

I'll leave this now as is and come back after all three of you have reviewed. Thanks for the help and guidance so far @ev-br.

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,47 +217,64 @@ def __init__(self, points, values, method="linear", bounds_error=True,
             self._validate_grid_dimensions(points, method)
         self.method = method
         self.bounds_error = bounds_error
+        self.grid, self.descending_dimensions = self._check_points(points)
```
```suggestion
        self.grid, self._descending_dimensions = self._check_points(points)
```

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,47 +217,64 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+        self.values = self._check_values(values)
+        self._check_dimensionality(self.grid, self.values)
+        self.fill_value = self._check_fill_value(self.values, fill_value)
+        if self.descending_dimensions:
```
```suggestion
        if self._descending_dimensions:
```

ev-br commented on scipy/interpolate/_rgi.py:
```diff
@@ -217,47 +217,64 @@ def __init__(self, points, values, method="linear", bounds_error=True,
+        self._check_dimensionality(self.grid, self.values)
+        self.fill_value = self._check_fill_value(self.values, fill_value)
+        if self.descending_dimensions:
+            self.values = np.flip(values, axis=self.descending_dimensions)
```
```suggestion
            self.values = np.flip(values, axis=self._descending_dimensions)
```

ev-br reviewed (commented):
`self.descending_dimensions` is an implementation detail, should get a leading undescore.

kmuehlbauer commented on benchmarks/benchmarks/interpolate.py:
```diff
@@ -249,30 +249,121 @@ class RegularGridInterpolator(Benchmark):
+        # check dimensionality
+        self._check_dimensionality(self.grid, values)
+        # flip, if needed
+        self.values = np.flip(values, axis=self.descending_dimensions)
```
Need to be changed here too.

ev-br replied:
Yes. Am going to squash-merge when the CI comes back green.

ev-br commented on benchmarks/benchmarks/interpolate.py:
```diff
@@ -249,30 +249,121 @@ class RegularGridInterpolator(Benchmark):
+        # check dimensionality
+        self._check_dimensionality(self.grid, values)
+        # flip, if needed
+        self.values = np.flip(values, axis=self.descending_dimensions)
```
```suggestion
        self.values = np.flip(values, axis=self._descending_dimensions)
```

ev-br commented:
Cirrus CI failures are unrelated and (are worked around in main anyway). Let's merge this to get the ball rolling. 
Thank you @egouden for starting this and @kmuehlbauer for making it across the finish line.

kmuehlbauer commented:
Thanks @ev-br for mentoring this PR and thanks @egouden for initial idea and setup.

egouden commented:
Thank you all for the nice collaborative work on this optimization.
